pennylane/devices/execution_config.py

```diff
@@ -30,16 +31,26 @@
 ]
 
 
-@dataclass
+@dataclass(frozen=True)
 class ExecutionConfig:
     """
     A class to configure the execution of a quantum circuit on a device.
 
     See the Attributes section to learn more about the various configurable options.
     """
 
-    shots: Optional[Union[int, Tuple[int]]] = None
-    """The number of shots for an execution"""
+    # pylint: disable=too-many-instance-attributes
+
+    shots: int = None
+    """The number of shots for an execution."""
+
+    shot_vector: List[ShotTuple] = None
+    """List of groupings of shots for an execution."""
```

Comment:
trying to think about how to handle this - it shouldn't be public (that is to say, set by users). Maybe dataclass can help us with aliasing? or at least we can make this field hidden to users/docs?

Reply:
I agree that this needs to change. Currently, even if the user sets the `shot_vector` in the constructor, its final value will be determined by `shots` due to the `__post_init__` which is good for the way `ExecutionConfig` will be used internally by PL. But it will be confusing for a user who creates an `ExecutionConfig` manually and expects the shot vector to be what they specified (assuming that the specified `shots` and `shot_vector` don't correspond with one another, which is likely to be the case if `shots` isn't in the constructor but `shot_vector` is).

Reply:
While old devices store `shot_vector` as a `Tuple[ShotTuple]`, we could store it as `Tuple[Int]` here. `Tuple[Int]` is the user-facing way of specifying a shot vector anyway.  Maybe we can have an additional property for the `Tuple[ShotTuple]` representation that is used internally.

Reply:
I thought the user would pass the shot-vector to `shots`, and it gets transformed internally, no? If we don't want it hidden, then this is perfect (with the updated type-hint). If we do want it hidden, instead of setting the value to `None` here, we can set it to `field(default=None, init=False, repr=False)` (plus, as mentioned in my comment in the other thread, this will make it invisible to Sphinx, for better _and_ for worse!).

Reply:
> While old devices store `shot_vector` as a `Tuple[ShotTuple]`, we could store it as `Tuple[Int]` here.

`_raw_shot_sequence` stores the `Tuple[int]` shot vector that the user passes to a device, and as far as I can tell `shot_vector` is a purely developer facing attribute (or at least is intended to be).

With that in mind, I like @timmysilv 's suggestion about setting `shot_vector` to `field(default=None, init=False, repr=False)` so that it is hidden from the user